File: src/main/java/app/application/controller/main_window/PlaylistPanelController.java
package app.application.controller.main_window;

import app.application.components.VideoElement;
import app.application.data.entities.YoutubePlaylist;
import app.application.exception.CantAbortDownloadException;
import app.application.exception.InvalidPlaylistUrlException;
import app.application.factories.VideoElementFactory;
import app.application.utils.DownloadExecutorHandler;
import app.application.utils.YoutubeIdExtractor;
import app.application.utils.YoutubeUrlValidator;
import app.application.utils.service.data.YoutubePlaylistDataService;
import app.application.utils.service.download.YoutubePlaylistDownloadService;
import javafx.beans.binding.Bindings;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.CheckBox;
import javafx.scene.control.Label;
import javafx.scene.control.ListView;
import javafx.scene.control.TextField;
import javafx.scene.layout.AnchorPane;
import org.springframework.stereotype.Component;

@Component
public class PlaylistPanelController {

	@FXML
	private AnchorPane playlistPanel;

	@FXML
	private TextField txtPlaylistLink;

	@FXML
	private ListView<VideoElement> listPlaylist;

	@FXML
	private Label txtPlaylistTitle;

	@FXML
	private Label lblDownloadProgress;

	@FXML
	private Button btnSearchPlaylist;

	@FXML
	private CheckBox chkAudioOnly;

	private final YoutubePlaylistDataService youtubePlaylistDataService;

	private final YoutubePlaylistDownloadService youtubePlaylistDownloadService;

	private final YoutubeIdExtractor youtubeIdExtractor;

	private final YoutubeUrlValidator youtubeUrlValidator;

	private final VideoElementFactory videoElementFactory;

	private DownloadExecutorHandler downloadExecutorHandler;

	public PlaylistPanelController(
					YoutubePlaylistDownloadService youtubePlaylistDownloadService,
					YoutubePlaylistDataService youtubePlaylistDataService,
					YoutubeIdExtractor youtubeIdExtractor,
					YoutubeUrlValidator youtubeUrlValidator,
					VideoElementFactory videoElementFactory) {
		this.youtubePlaylistDownloadService = youtubePlaylistDownloadService;
		this.youtubePlaylistDataService = youtubePlaylistDataService;
		this.youtubeIdExtractor = youtubeIdExtractor;
		this.youtubeUrlValidator = youtubeUrlValidator;
		this.videoElementFactory = videoElementFactory;
	}

	private YoutubePlaylist youtubePlaylist;

	@FXML
	public void initialize(){
		youtubePlaylistDownloadService.setLabel(lblDownloadProgress);
		btnSearchPlaylist.disableProperty().bind(Bindings.isEmpty(txtPlaylistLink.textProperty()));
		downloadExecutorHandler = new DownloadExecutorHandler();
	}

	public void btnSearchPlaylistClick() throws InvalidPlaylistUrlException {
		listPlaylist.getItems().clear();
		youtubeUrlValidator.checkPlaylistUrl(txtPlaylistLink.getText());
		youtubePlaylist = youtubePlaylistDataService.getPlaylistInfo(youtubeIdExtractor.getPlayListIdFromLink(txtPlaylistLink.getText()));
		lblDownloadProgress.setText("Videos: 0/" + youtubePlaylist.getPlaylistSize());
		txtPlaylistTitle.setText(youtubePlaylist.getPlaylistTitle());
		youtubePlaylist.getPlaylistVideos().forEach(video ->
				listPlaylist.getItems().add(videoElementFactory.createVideoElement(video))
		);
		chkAudioOnly.selectedProperty().bindBidirectional(youtubePlaylist.getAudioOnly());
		playlistPanel.setVisible(true);
	}

	public void btnPlaylistDownloadClick(){
		downloadExecutorHandler.executeTask(() -> youtubePlaylistDownloadService.downloadPlaylist(youtubePlaylist));
	}

	public void btnAbortClick() throws CantAbortDownloadException {
		downloadExecutorHandler.killTask();
	}

}

File: src/main/kotlin/app/application/controller/main_window/PlaylistPanelController.kt
package app.application.controller.main_window

import app.application.components.VideoElement
import app.application.data.entities.YoutubePlaylist
import app.application.data.entities.YoutubePlaylistVideoDetail
import app.application.exception.CantAbortDownloadException
import app.application.exception.InvalidPlaylistUrlException
import app.application.spring.factories.VideoElementFactory
import app.application.utils.DownloadExecutorHandler
import app.application.utils.DownloadExecutorHandler.DownloaderTask
import app.application.spring.service.YoutubeIdExtractor
import app.application.spring.service.YoutubeUrlValidator
import app.application.spring.service.data.YoutubePlaylistDataService
import app.application.spring.service.download.YoutubePlaylistDownloadService
import javafx.beans.binding.Bindings
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.AnchorPane
import org.springframework.stereotype.Component
import java.util.function.Consumer

@Component
class PlaylistPanelController(
    private val youtubePlaylistDownloadService: YoutubePlaylistDownloadService,
    private val youtubePlaylistDataService: YoutubePlaylistDataService,
    private val youtubeIdExtractor: YoutubeIdExtractor,
    private val youtubeUrlValidator: YoutubeUrlValidator,
    private val videoElementFactory: VideoElementFactory
) {

    @FXML
    private lateinit var playlistPanel: AnchorPane

    @FXML
    private lateinit var txtPlaylistLink: TextField

    @FXML
    private lateinit var listPlaylist: ListView<VideoElement>

    @FXML
    private lateinit var txtPlaylistTitle: Label

    @FXML
    private lateinit var lblDownloadProgress: Label

    @FXML
    private lateinit var btnLoadPlaylist: Button

    @FXML
    private lateinit var chkAudioOnly: CheckBox

    private val downloadExecutorHandler: DownloadExecutorHandler = DownloadExecutorHandler()
    private var youtubePlaylist: YoutubePlaylist? = null

    @FXML
    fun initialize() {
        youtubePlaylistDownloadService.label = lblDownloadProgress
        btnLoadPlaylist.disableProperty().bind(Bindings.isEmpty(txtPlaylistLink.textProperty()))
    }

    @Throws(InvalidPlaylistUrlException::class)
    fun btnLoadPlaylistClick() {
        val url = txtPlaylistLink.text.trim()
        listPlaylist.items.clear()
        youtubeUrlValidator.checkPlaylistUrl(url)
        youtubePlaylist = youtubePlaylistDataService.getPlaylistInfo(youtubeIdExtractor.getPlayListIdFromLink(url))
        lblDownloadProgress.text = "Videos: 0/" + youtubePlaylist?.playlistSize
        txtPlaylistTitle.text = youtubePlaylist?.playlistTitle
        youtubePlaylist?.playlistVideos?.forEach(Consumer { video: YoutubePlaylistVideoDetail? -> listPlaylist.items.add(videoElementFactory.createVideoElement(video)) })
        chkAudioOnly.selectedProperty().bindBidirectional(youtubePlaylist?.audioOnly)
        playlistPanel.isVisible = true
    }

    fun btnPlaylistDownloadClick() {
        downloadExecutorHandler.executeTask(object : DownloaderTask {
            override fun execute() {
                youtubePlaylist?.let { youtubePlaylistDownloadService.downloadPlaylist(it) }
            }
        })
    }

    fun loadKeyPressed(keyEvent: KeyEvent){
        if(keyEvent.code == KeyCode.ENTER){
            btnLoadPlaylistClick()
        }
    }

    @Throws(CantAbortDownloadException::class)
    fun btnAbortClick() {
        downloadExecutorHandler.killTask()
    }
}